### DISCUSSION TITLE
Help with AWS IOT

### DISCUSSION THREAD

**Discussion by g0lgs** [arendst/Tasmota#12642] [Q&A] [answered]
I am investigating using AWS IoT with some Tasmota Devices, but I am stuck when trying to follow the instructions at:

https://tasmota.github.io/docs/AWS-IoT/#4-configure-aws-iot-policy-to-be-done-once

I do NOT see the Launch Stack or Create Stack on AWS console.

can anyone point me in the right direction ?

**Comment by s-hadinger** [comment, marked as answer]
You need to select the service "CloudFormation". But first choose Tha appropriate AWS région. Can you access it?

**Reply by g0lgs** [reply]
Ok so it seems I was looking in the wrong place as I was looking under AWS IoT.

I will need to fathom out how using CloudFormation affects what I'm already doing in AWS IoT and how that affects the costs.

**Reply by s-hadinger** [reply]
CloudFormation per se does not cost anything. There is a lambda function used as authorizer but the result is cached so the cost impact should be zero

**Reply by g0lgs** [reply]
I have compiled Tasmota, using Visual Studio Code and platformio on Ubuntu 20.04 (as I have done fine for using my own MQTT host) as per the instructions presetting Wifi SSID's and Passwords. It connects to WiFi fine.

I completed the AWS CloudFormation bit and I got a set of commands to paste into the console, but that seems to trigger some odd actions where the device keeps restarting and reconnecting to the Wifi and flashing some of my relays and LEDs and I can no longer access it's Web interface.

So I put the same build of Tasmota onto an ESP8266 module via USB (using esptool on Linux machine) leaving the default BASIC GPIO assignments and repeated the MQTT configuration (using picocom) I get:

```
08:43:13.698 RSL: RESULT = {"SetOption3":"ON"}
08:43:13.951 RSL: RESULT = {"SetOption103":"ON"}
08:43:14.201 RSL: RESULT = {"MqttHost":"xxxxxxxxxxxxxxxx.iot.eu-west-2.amazonaws.com"}
08:43:14.451 RSL: RESULT = {"MqttPort":443}
08:43:14.702 RSL: RESULT = {"MqttUser":"tasmota?x-amz-customauthorizer-name=TasmotaAuth"}
08:43:14.906 RSL: RESULT = {"MqttPassword":"xxxxxxxxxxxx/xxxxxxxxxxx=="}
08:43:15.050 MQT: Attempting connection...

```
User exception (panic/abort/assert)
--------------- CUT HERE FOR EXCEPTION DECODER ---------------

Abort called

This is followed by a whole load of hex codes and then it restarts showing the following:

```
00:00:00.001 HDW: ESP8266EX
00:00:00.042 CFG: Loaded from flash at F6, Count 6
00:00:00.047 QPC: Count 1
00:00:00.056 Project tasmota Tasmota Version 9.5.0.2(tasmota)-2_7_4_9(2021-07-15T08:39:18)
00:00:00.544 WIF: Connecting to AP1 XXXXXXXXXXXXX Channel 6 BSSId XX:XX:XX:XX:XX:XX in mode 11n as tasmota_B07E0F-7695...
00:00:02.754 WIF: Connected
00:00:03.005 HTP: Web server active on tasmota_B07E0F-7695 with IP address 192.168.xx.xxx
08:43:27.008 RSL: INFO1 = {"Info1":{"Module":"Sonoff Basic","Version":"9.5.0.2(tasmota)","FallbackTopic":"cmnd/DVES_B07E0F_fb/","GroupTopic":"cmnd/tasmotas/"}}
08:43:27.013 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota_B07E0F-7695","IPAddress":"192.168.80.252"}}
08:43:27.024 RSL: INFO3 = {"Info3":{"RestartReason":{"Exception":4,"Reason":"Software Watchdog","EPC":["4000e289","00000000","00000000"],"EXCVADDR":"00000000","DEPC":"00000000"}}}
08:43:27.039 RSL: RESULT = {"POWER":"OFF"}
08:43:27.042 RSL: POWER = OFF
08:43:29.445 QPC: Reset
08:43:31.416 RSL: STATE = {"Time":"2021-07-15T08:43:31","Uptime":"0T00:00:10","UptimeSec":10,"Heap":22,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER":"OFF","Wifi":{"AP":1,"SSId":"BTHub3-C3D2","BSSId":"74:83:C2:21:C2:61","Channel":6,"Mode":"11n","RSSI":76,"Signal":-62,"LinkCount":1,"Downtime":"0T00:00:04"}}
08:48:31.418 RSL: STATE = {"Time":"2021-07-15T08:48:31","Uptime":"0T00:05:10","UptimeSec":310,"Heap":20,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER":"OFF","Wifi":{"AP":1,"SSId":"BTHub3-C3D2","BSSId":"74:83:C2:21:C2:61","Channel":6,"Mode":"11n","RSSI":76,"Signal":-62,"LinkCount":1,"Downtime":"0T00:00:04"}}

```

**Reply by s-hadinger** [reply]
Can you please try flashing tasmota-zbbridge pre-compiled binary? This one is known to work with AWS IOT.

**Reply by g0lgs** [reply]
Ok so using the tasmota-zbbridge pre-compiled binary allows both my ESP8266 module and my device to run and publish to AWS IoT.

So the puzzle is why does my own built image not work

**Reply by s-hadinger** [reply]
Good. What options did you change?

**Reply by g0lgs** [reply]
My user_config_override. h looks like:

```
#ifndef _USER_CONFIG_OVERRIDE_H_
#define _USER_CONFIG_OVERRIDE_H_

#undef  STA_SSID1
#define STA_SSID1         "xxxxxxxxxxxx"             // [Ssid1] Wifi SSID

#undef  STA_PASS1
#define STA_PASS1         "xxxxxxxxxxx"     // [Password1] Wifi password

#undef  STA_SSID2
#define STA_SSID2         "xxxxxxxxxxxx"             // [Ssid2] Wifi SSID

#undef  STA_PASS2
#define STA_PASS2         "xxxxxxxxxxxx"     // [Password2] Wifi password

#ifndef USE_MQTT_TLS 
#define USE_MQTT_TLS
#define USE_MQTT_TLS_CA_CERT
#endif
#ifndef USE_MQTT_AWS_IOT_LIGHT
#define USE_MQTT_AWS_IOT_LIGHT
#endif
#ifdef USE_DISCOVERY
#undef USE_DISCOVERY
#endif

#endif  // _USER_CONFIG_OVERRIDE_H_
```

**Reply by s-hadinger** [reply]
I don't see anything wrong here. I can't test until next week. Is the zbbridge firmware enough for your needs until then?

**Reply by s-hadinger** [reply]
Can you try with ` #define USE_MQTT_TLS_FORCE_EC_CIPHER`

**Reply by g0lgs** [reply]
So adding `#define USE_MQTT_TLS_FORCE_EC_CIPHER` appears to have worked for my ESP8266 Module.

However somehow my Power Strip device that I earlier tested the zbridge version on now will not stay on the WiFi network (either on my real WiFi or in access point mode) long enough for me to try to re-flash it or do anything much else with it - as it has anti-tamper screws in the case and the only way I managed to flash it in the first place was OTA that is a bit of a problem.

It stays up on the real Wifi (after putting the credentials in in AP mode) just about long enough for me to get to the Web Console to see:

```
00:00:00.001 HDW: ESP8285
00:00:00.025 CFG: Loaded from flash at FA, Count 5
00:00:00.028 QPC: Count 1
00:00:00.032 SNS: Hardware Serial
00:00:00.034 Project tasmota Tasmota Version 9.5.0(zbbridge)-2_7_4_9(2021-06-17T08:27:09)
00:00:00.035 ZIG: Resetting EZSP device
00:00:00.619 WIF: Connecting to AP1 XXXXXXXX in mode 11n as tasmota_4435CF-5583...
00:00:05.003 WIF: Connected
00:00:05.254 HTP: Web server active on tasmota_4435CF-5583 with IP address 192.168.12.202
00:00:06.536 QPC: Reset
00:00:09.840 RSL: INFO1 = {"Info1":{"Module":"Sonoff ZbBridge","Version":"9.5.0(zbbridge)","FallbackTopic":"cmnd/DVES_4435CF_fb/","GroupTopic":"cmnd/tasmotas/"}}
00:00:09.842 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota_4435CF-5583","IPAddress":"192.168.12.202"}}
00:00:09.842 RSL: INFO3 = {"Info3":{"RestartReason":"Software/System restart"}}
00:00:14.070 RSL: STATE = {"Time":"1970-01-01T00:00:14","Uptime":"0T00:00:15","UptimeSec":15,"Vcc":3.466,"Heap":29,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":736,"MqttCount":0,"Wifi":{"AP":1,"SSId":"EE-Test","BSSId":"64:70:02:20:9C:5E","Channel":1,"Mode":"11n","RSSI":72,"Signal":-64,"LinkCount":1,"Downtime":"0T00:00:06"}}
00:00:16.140 ZIG: Resetting EZSP device
15:11:22.182 ZIG: timeout, goto label 99
15:11:22.184 RSL: RESULT = {"ZbState":{"Status":99,"Message":"Abort"}}
15:11:22.185 ZIG: Abort
15:11:22.185 ZIG: Stopping (99)

```
Just after that it reboots and appears in AP mode again.

Is there something I can do that might stop it restarting so frequently ?

I just get as far as trying to upload tasmota-minimal.bin.gz , but even that failed with a message about no room and then it rebooted.

**Reply by s-hadinger** [reply]
If you have the console, try `so3 0` as it will disable Mqtt. Otherwise cut the internet access on your router. The crash happens during the TLS handshake

**Reply by g0lgs** [reply]
I managed to do `so3 0` but it still keeps restarting and setting `so3` back to ON.

**Reply by Jason2866** [reply]
Have you tried to cut Internet access of your router?

**Reply by g0lgs** [reply]
No (in part because it was the Office WiFi I was using and that was not an option) - it is too late for that now as the one Power Strip is no longer working at all - after I tried various other things - it does not do anything apart from turn one outlet on when power is applied (no tasmota_xxxxx AP and not connecting to the WiFi)- I will try to get the screws out and wire it up to re-flash but that will be a few days now.

I do get the repeated resets/restarts with just the seperate ESP-12F module and the zbridge version too now, but not fathom'd out why (even tried erasing the flash and re-flashing the downloaded firmware and connecting to the WiFi without doing the AWS config).

**Reply by g0lgs** [reply]
After a bit more fiddling it appears that if you erase the flash and directly flash the zbridge version onto a ESP-12F it enters this continual reset/restart loop, if you load one of the other versions (I just used tasmota.bin) and connect it to the WiFi and then 'upgrade' to the zbridge version (might have to be via the minimal version) all is fine - unless you then do the `reset 3` command at which point it enters the reset/restart loop.

**Reply by s-hadinger** [reply]
Are you using a private key? Otherwise Reset 3 only clears wifi calibration.

**Reply by g0lgs** [reply]
only WPA-PSK

**Reply by s-hadinger** [reply]
I mean how do you authenticate to AWS IOT? Private key or password?

**Reply by g0lgs** [reply]
Using password.

**Reply by s-hadinger** [reply]
Then it doesn't make sense that Reset 3 creates a bootloop. Do you have logs?

**Reply by g0lgs** [reply]
If I have some time in the next couple of days I will try to get logs showing what happens when I tried reset 3.

However using my ESP-12F with esptool.py V3.1 on Ubuntu 20.04

Erasing the flash and then flashing the zbridge version, the best I can get is getting my (other) desktop OS to keep reconnecting to the device in AP mode and open `http://192.168.4.1/cs?`

(SerialLog is defaulting to off and the reset keeps going back to that even if I manage to change it to 2 from the Web Console - so I have have been unable tp enter any USB Serial console commands).

So this is what I see in the web console (if I do not try to enter WiFi details and just leave it):

```
00:00:00.001 HDW: ESP8266EX
00:00:00.018 CFG: Loaded from flash at 1FB, Count 1
00:00:00.021 QPC: Count 1
00:00:00.024 ZIG: Randomizing Zigbee parameters, please check with 'ZbConfig'
00:00:00.026 SNS: Hardware Serial
00:00:00.027 Project tasmota Tasmota Version 9.5.0(zbbridge)-2_7_4_9(2021-06-17T08:27:09)
00:00:00.028 ZIG: Resetting EZSP device
00:00:00.230 WIF: WifiManager active for 3 minutes
00:00:00.736 HTP: Web server active on tasmota_B07E0F-7695 with IP address 192.168.4.1
00:00:06.988 QPC: Reset
00:00:15.507 ZIG: Resetting EZSP device
00:00:20.673 ZIG: timeout, goto label 99
00:00:20.675 RSL: RESULT = {"ZbState":{"Status":99,"Message":"Abort"}}
00:00:20.675 ZIG: Abort
00:00:20.675 ZIG: Stopping (99)
```
Whilst this appears on the Serial:

```
00:00:00.001 HDW: ESP8266EX
00:00:00.018 CFG: Loaded from flash at 1FB, Count 1
00:00:44.504 CFG: Use defaults

 ets Jan  8 2013,rst cause:1, boot mode:(3,7)

load 0x4010f000, len 3584, room 16 
tail 0
chksum 0xb0
csum 0xb0
v3969889e
~ld
```

**The whole of that takes about 45 seconds.**

This is from the Web Console after I manage to enter WiFi details and the browser gets redirected to the WiFi IP:
```

00:00:00.001 HDW: ESP8266EX
00:00:00.026 CFG: Loaded from flash at FA, Count 5
00:00:00.029 QPC: Count 1
00:00:00.034 SNS: Hardware Serial
00:00:00.035 Project tasmota Tasmota Version 9.5.0(zbbridge)-2_7_4_9(2021-06-17T08:27:09)
00:00:00.036 ZIG: Resetting EZSP device
00:00:00.570 WIF: Connecting to AP1 XXXXXXXXX in mode 11n as tasmota_B07E0F-7695...
00:00:06.689 QPC: Reset
00:00:07.009 WIF: Connected
00:00:07.261 HTP: Web server active on tasmota_B07E0F-7695 with IP address 10.44.100.144
23:04:40.281 RSL: INFO1 = {"Info1":{"Module":"Sonoff ZbBridge","Version":"9.5.0(zbbridge)","FallbackTopic":"cmnd/DVES_B07E0F_fb/","GroupTopic":"cmnd/tasmotas/"}}
23:04:40.282 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota_B07E0F-7695","IPAddress":"10.44.100.144"}}
23:04:40.283 RSL: INFO3 = {"Info3":{"RestartReason":"Software/System restart"}}
23:04:44.654 RSL: STATE = {"Time":"2021-07-16T23:04:44","Uptime":"0T00:00:14","UptimeSec":14,"Vcc":3.439,"Heap":30,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Wifi": {"AP":1,"SSId":"XXXXXXXXXX","BSSId":"AA:BB:CC:DD:EE:FF","Channel":11,"Mode":"11n","RSSI":64,"Signal":-68,"LinkCount":1,"Downtime":"0T00:00:08"}}
23:04:47.254 ZIG: Resetting EZSP device
23:04:52.430 ZIG: timeout, goto label 99
23:04:52.432 RSL: RESULT = {"ZbState":{"Status":99,"Message":"Abort"}}
23:04:52.433 ZIG: Abort
23:04:52.433 ZIG: Stopping (99)
23:05:16.851 RSL: RESULT = {"Reset":"Reset and Restarting"}
23:05:17.798 CFG: Use defaults
```
That whole process probably takes a little longer, than the 45secs between the rebbot when I do nothing, but it seems much less than 3 minutes.

At which point it goes back to AP mode.

**Reply by s-hadinger** [reply]
Something is wrong. Why is there a reset sent at 23:05:16 ?

**Reply by g0lgs** [reply]
> Something is wrong. Why is there a reset sent at 23:05:16 ?

Nothing that I'm doing.

**Reply by s-hadinger** [reply]
I will try tomorrow with a Wemos D1. But I don't see where this reset comes from.

**Reply by s-hadinger** [reply]
In between can toy set seriallog 4? I want to know where this reset comes from with serial logging

**Reply by g0lgs** [reply]
SerialLog 4 (on USB) after it connected to WiFi looks like:

```
00:00:00.001 HDW: ESP8266EX
00:00:00.029 CFG: Loaded from flash at F9, Count 6
00:00:06.746 RSL: RESULT = {"SerialLog":{"4":{"Active":"4"}}}
00:00:06.890 CFG: Saved to flash at F8, Count 7, Bytes 4096
00:00:06.923 QPC: Reset
00:00:07.044 RTC: UTC 2021-07-17T10:00:16, DST 2021-03-28T02:00:00, STD 2021-10-31T03:00:00
11:00:16.005 HTP: Main Menu
11:00:16.315 RSL: INFO1 = {"Info1":{"Module":"Sonoff ZbBridge","Version":"9.5.0(zbbridge)","FallbackTopic":"cmnd/DVES_B07E0F_fb/","GroupTopic":"cmnd/tasmotas/"}}
11:00:16.319 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota_B07E0F-7695","IPAddress":"10.44.100.144"}}
11:00:16.330 RSL: INFO3 = {"Info3":{"RestartReason":"Software/System restart"}}
11:00:16.861 CFG: Saved to flash at F7, Count 8, Bytes 4096
11:00:17.707 APP: Boot Count 3
11:00:17.900 CFG: Saved to flash at 1FB, Count 9, Bytes 4096
11:00:20.723 RSL: STATE = {"Time":"2021-07-17T11:00:20","Uptime":"0T00:00:12","UptimeSec":12,"Vcc":3.439,"Heap":30,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Wifi":{"AP":1,"SSId":"XXXXXXXXX","BSSId":"AA:BB:CC:DD:EE:FF","Channel":11,"Mode":"11n","RSSI":74,"Signal":-63,"LinkCount":1,"Downtime":"0T00:00:06"}}
11:00:24.310 ZIG: Resetting EZSP device
11:00:29.471 ZIG: timeout, goto label 99
11:00:29.473 RSL: RESULT = {"ZbState":{"Status":99,"Message":"Abort"}}
11:00:29.474 ZIG: Abort
11:00:29.474 ZIG: Stopping (99)
11:00:35.119 WIF: Checking connection...
11:00:55.125 WIF: Checking connection...
11:00:56.027 SRC: Button
11:00:56.028 CMD: Grp 0, Cmnd 'RESET', Idx 1, Len 1, Data '1'
11:00:56.030 RSL: RESULT = {"Reset":"Reset and Restarting"}
11:00:56.878 CFG: Use defaults
11:00:56.880 CFG: CR 96/699, Busy 0
11:00:56.880 CFG: CR 103/699, Busy 0
11:00:56.880 CFG: CR 111/699, Busy 0
11:00:56.880 CFG: CR 119/699, Busy 0
11:00:56.884 CFG: CR 127/699, Busy 0
11:00:56.887 CFG: CR 134/699, Busy 0
11:00:56.890 CFG: CR 187/699, Busy 0
11:00:56.894 CFG: CR 187/699, Busy 0
11:00:56.897 CFG: CR 187/699, Busy 0
11:00:56.900 CFG: CR 187/699, Busy 0
11:00:56.903 CFG: CR 187/699, Busy 0
11:00:56.907 CFG: CR 194/699, Busy 0
11:00:56.910 CFG: CR 194/699, Busy 0
11:00:56.913 CFG: CR 194/699, Busy 0
11:00:56.917 CFG: CR 194/699, Busy 0
11:00:56.920 CFG: CR 194/699, Busy 0
11:00:56.923 CFG: CR 203/699, Busy 0
11:00:56.927 CFG: CR 212/699, Busy 0
11:00:56.930 CFG: CR 221/699, Busy 0
11:00:56.933 CFG: CR 233/699, Busy 0
11:00:56.936 CFG: CR 234/699, Busy 0
11:00:56.940 CFG: CR 235/699, Busy 0
11:00:56.943 CFG: CR 243/699, Busy 0
11:00:56.946 CFG: CR 260/699, Busy 0
11:00:56.950 CFG: CR 264/699, Busy 0
11:00:56.953 CFG: CR 268/699, Busy 0
11:00:56.956 CFG: CR 272/699, Busy 0
11:00:56.959 CFG: CR 275/699, Busy 0
11:00:56.963 CFG: CR 277/699, Busy 0
11:00:56.966 CFG: CR 283/699, Busy 0
11:00:56.969 CFG: CR 287/699, Busy 0
11:00:56.973 CFG: CR 299/699, Busy 0
11:00:56.976 CFG: CR 314/699, Busy 0
11:00:56.979 CFG: CR 331/699, Busy 0
11:00:56.983 CFG: CR 331/699, Busy 0
11:00:56.986 CFG: CR 331/699, Busy 0
11:00:56.989 CFG: CR 331/699, Busy 0
11:00:57.184 CFG: Saved to flash at 1FB, Count 1, Bytes 4096
11:00:57.832 CFG: Saved to flash at FD, Count 2, Bytes 4096
11:00:58.872 APP: Restarting

 ets Jan  8 2013,rst cause:1, boot mode:(3,7)

load 0x4010f000, len 3584, room 16 
tail 0
chksum 0xb0
csum 0xb0
v3969889e
~ld
```

Just to be clear this is a ESP-12F mouted on a 'ESP-Downloader' with no connections to anything else.

Info for the Downloader is at `https://github.com/linshuqin329/ESP-Downloader/tree/master/Downloader_for_ESP8266-Modules`

**Reply by s-hadinger** [reply]
Ah got it. It thinks that Button 1 is pressed for more than 40 s which triggers a factory reset. Use `template 1` as soon as you can

**Reply by s-hadinger** [reply]
Sorry I meant `module 1`

**Reply by ascillato2** [reply]
Hi,

I have run the same configuration as @g0lgs (following also the guide https://tasmota.github.io/docs/AWS-IoT/ ) and without the `#define USE_MQTT_TLS_FORCE_EC_CIPHER`, the ESP8266 crashed while trying to connect.

The configuration that works fine is:

```cpp
#ifndef USE_MQTT_TLS 
#define USE_MQTT_TLS
#define USE_MQTT_TLS_CA_CERT
#define USE_MQTT_TLS_FORCE_EC_CIPHER
#endif
#ifndef USE_MQTT_AWS_IOT_LIGHT
#define USE_MQTT_AWS_IOT_LIGHT
#endif
#ifdef USE_DISCOVERY
#undef USE_DISCOVERY
#endif
```

So, @s-hadinger what do you think? we should add `#define USE_MQTT_TLS_FORCE_EC_CIPHER` to the docs? With that it works very good and very fast. (either in 80Mhz or 160Mhz the connecting and the responses are less than a second)

**Reply by g0lgs** [reply]
> Sorry I meant `module 1`

That seems to have stopped the restarting on my ESP8266.

I will need to see if I can re-flash the Tuya TYWE2S module from the Power Strip (in the Office) and try to get that working again (having manged to completely stop it working 2 days ago).

**Reply by ascillato** [reply]
With the fix https://github.com/arendst/Tasmota/pull/12687 the key `#define USE_MQTT_TLS_FORCE_EC_CIPHER` is not needed anymore.

Thanks @s-hadinger

**Reply by g0lgs** [reply]
With a bit more care than previously with the power strip (using hand tools instead of battery powered screwdriver) I managed to get the screws out of the power strip and have re-programmed the TYWE2S module.
